zkevm-circuits/src/circuit_tools/cell_manager.rs

```diff
@@ -236,12 +236,25 @@ impl<F: Field, C: CellType> Expr<F> for CellColumn<F, C> {
     }
 }
 
+
 #[derive(Clone, Debug)]
 pub struct CellManager<F, C: CellType> {
     configs: Vec<CellConfig<C>>,
     columns: Vec<CellColumn<F, C>>,
     height: usize,
+    width: usize,
     height_limit: usize,
+
+    // branch ctxs
+    branch_ctxs: HashMap<String, CmContext<F, C>>,
```

Comment:
Why is it a hashmap and not a simple vector? Contexts will work like a stack no? Similarly to how we handle the condition in the constraint builder. So if you enter a context you just push a new context on the stack, if you exit a context you just pop it from the stack?

Reply:
https://github.com/Brechtpd/zkevm-circuits/pull/9 If you scrolled down to one of my replies:

> I thought of it and figured that a **dynamic branching callstack** is equivalent to a s**tatic upside-down tree**:
> 
> ```
>    |_if_| |_else_|     |_if_|
>   |_____if_______|    |_else_|
>  |__________________________|
> 
> ```
> A simple callstack structure won't work cuz we need to remember what happened in the past. Thus we have to save the entire tree.

In the condition case remember we had to use `region_id` to hack around this problem, basically manually divide the branching callstack into states labeled with id:
```
   |_if_| |_else_|     |_if_|
  |___ match____|    |_match__|
 |______________________________|
 (      START     )(   ACCOUNT   ) ....
```
And one time i said ideally we can make store expressions work like CM ctx by changing the macros, but I failed 😭. And that could be something we can do for the next iteration.